hooks/useTransactionSender.tsx

```diff
@@ -1,25 +1,320 @@
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-yellow-filled)"
+        />
+      );
+    }
+
+    return <Loader size="xs" color="var(--mantine-color-dark-2)" />;
+  };
+
+  /**
+   * asynchronously confirm transaction result for each signature and update state when resolved
+   */
+  const confirmTransaction = async (transactionInfos: Array<TransactionInfo<T>>, id: string) => {
+    transactionInfos.forEach(async (transactionInfo) => {
+      const controller = new AbortController();
+      const signal = controller.signal;
+
+      const { blockhash, lastValidBlockHeight } = await connection.getLatestBlockhash();
+      const confirmTransactionPromise = connection.confirmTransaction({
+        signature: transactionInfo.signature,
+        blockhash,
+        lastValidBlockHeight,
+      });
+
+      const timeoutPromise = new Promise((_, reject) => {
+        /**
+         * note: we use a constant timeout of 30s (seems standard across many solana projects) for simplicity
+         * for now. in the future, we might opt to dynamically set this timeout based on various cluster stats,
+         * like ping time
+         */
+        const timeoutId = setTimeout(() => {
+          controller.abort();
+        }, 30_000);
+
+        signal.addEventListener('abort', () => {
+          clearTimeout(timeoutId);
+          reject(new Error('Transaction confirmation aborted'));
+        });
+      });
+
+      Promise.race([confirmTransactionPromise, timeoutPromise])
+        .then((result) => {
+          if ((result as any).value) {
+            const transactionHasError =
+              (result as RpcResponseAndContext<SignatureResult>).value.err !== null;
+
+            updateSingleTransactionInfo(transactionInfo.signature, id, () => ({
+              result: {
+                err: transactionHasError ? {} : null,
+              },
+            }));
+
+            setSuccessfulSignatureCount((state) => {
+              const current = state[id] ?? 0;
+              return {
+                ...state,
+                [id]: transactionHasError ? current : current + 1,
+              };
+            });
+          }
+        })
+        .catch((err) => {
+          console.error(
+            `an error occurred resolving transaction for signature [${transactionInfo.signature}]`,
+            err,
+          );
+
+          updateSingleTransactionInfo(transactionInfo.signature, id, () => ({
+            status: TransactionStatus.TIMED_OUT,
+          }));
+        });
+    });
+  };
 
   const send = useCallback(
     /**
      * Sends transactions.
      * @param txs A sequence of sets of transactions. Sets are executed simultaneously.
      * @returns A sequence of set of tx signatures.
      */
-    async <T extends Transaction | VersionedTransaction>(txs: SingleOrArray<T>[]) => {
+    async (txs: SingleOrArray<T | TransactionWithMetadata<T>>[]) => {
```

Comment:
this allows us to modify existing `send` invocations to add some extra metadata related to the transaction, such as a name/description. i didn't include those here to keep the scope of changes small. however, i can easily add in another PR if that's something of interest.